src/Razor/src/Microsoft.AspNetCore.Razor.VSCode/src/RazorRenameFeature.ts

```diff
@@ -0,0 +1,96 @@
+import { RazorLogger } from './RazorLogger';
+import { RazorRenameProvider } from './RazorRenameProvider';
+import { LanguageKind } from './RPC/LanguageKind';
+
+export class RazorRenameFeature {
+    public readonly serviceClient: RazorLanguageServiceClient;
+    public readonly renameProvider: RazorRenameProvider;
+
+    constructor(
+        documentSynchronizer: RazorDocumentSynchronizer,
+        documentManager: RazorDocumentManager,
+        serviceClient: RazorLanguageServiceClient,
+        logger: RazorLogger) {
+            this.serviceClient = serviceClient;
+            this.renameProvider = new RazorRenameProvider(
+                documentSynchronizer,
+                documentManager,
+                serviceClient,
+                logger);
+    }
+
+    public register() {
+        const registrations = [
+            vscode.languages.registerRenameProvider(RazorLanguage.id, this.renameProvider),
+            vscode.commands.registerCommand(
+                'razor.remapRenameEdit', async (
+                    document: vscode.TextDocument,
+                    position: vscode.Position,
+                    newText: string,
+                    cSharpEdit: vscode.WorkspaceEdit) => {
+                    const remappedEdit = await this.remapRazorCSharpEdits(
+                    cSharpEdit,
+                    this.serviceClient);
+                    return remappedEdit;
+                }),
+        ];
+
+        return vscode.Disposable.from(...registrations);
+    }
+
+    // Given a set of C# edits, we need to map edits to Razor CSharp documents back to the original Razor documents.
+    private async remapRazorCSharpEdits(
+        cSharpEdit: vscode.WorkspaceEdit,
+        serviceClient: RazorLanguageServiceClient) {
+
+        const result = new vscode.WorkspaceEdit();
+
+        // The returned edits will be for the projected C# documents. We now need to re-map that to the original document.
+        for (const entry of cSharpEdit.entries()) {
+            const uri = entry[0];
+            const edits = entry[1];
+
+            if (!isRazorCSharpFile(uri)) {
+                // This edit happens outside of a Razor document. Let the edit go through as is.
+                result.set(uri, edits);
+                continue;
+            }
+
+            // We're now working with a Razor CSharp document.
+            const documentUri = getRazorDocumentUri(uri);
+            const newEdits = new Array<vscode.TextEdit>();
+
+            // Re-map each edit to its position in the original Razor document.
+            for (const edit of edits) {
+                const remappedResponse = await serviceClient.mapToDocumentRange(
+                    LanguageKind.CSharp,
+                    edit.range,
+                    documentUri);
+
+                if (!remappedResponse) {
+                    // Something went wrong when re-mapping to the original document. Accept the edit as is.
```

Comment:
Hmm do we want that? VSCode will pop up an error stating its innability to edit the unknown document. It might make sense just to throw this edit away.

Reply:
Throw the entire edit away or just the edit for this file? We probably don't want inconsistent renames but I'm not sure.

Reply:
Just the edit for this file.